src/unix/lwt_preemptive.ml

```diff
@@ -12,173 +11,106 @@
-let max_thread_queued = ref 1000
+let max_domains_queued = ref 128
 
 let get_max_number_of_threads_queued _ =
-  !max_thread_queued
+  !max_domains_queued
 
 let set_max_number_of_threads_queued n =
   if n < 0 then invalid_arg "Lwt_preemptive.set_max_number_of_threads_queued";
-  max_thread_queued := n
-
-(* The total number of preemptive threads currently running: *)
-let threads_count = ref 0
-
-(* +-----------------------------------------------------------------+
-   | Preemptive threads management                                   |
-   +-----------------------------------------------------------------+ *)
-
-module CELL :
-sig
-  type 'a t
-
-  val make : unit -> 'a t
-  val get : 'a t -> 'a
-  val set : 'a t -> 'a -> unit
-end =
-struct
-  type 'a t = {
-    m  : Mutex.t;
-    cv : Condition.t;
-    mutable cell : 'a option;
-  }
-
-  let make () = { m = Mutex.create (); cv = Condition.create (); cell = None }
-
-  let get t =
-    let rec await_value t =
-      match t.cell with
-      | None ->
-        Condition.wait t.cv t.m;
-        await_value t
-      | Some v ->
-        t.cell <- None;
-        Mutex.unlock t.m;
-        v
-    in
-    Mutex.lock t.m;
-    await_value t
+  max_domains_queued := n
 
-  let set t v =
-    Mutex.lock t.m;
-    t.cell <- Some v;
-    Mutex.unlock t.m;
-    Condition.signal t.cv
-end
+let domains_count = ref 0
 
-type thread = {
-  task_cell: (int * (unit -> unit)) CELL.t;
-  (* Channel used to communicate notification id and tasks to the
-     worker thread. *)
+type task =
+  Task of (int * (unit -> unit))
+  | Quit 
 
-  mutable thread : Thread.t;
-  (* The worker thread. *)
-
-  mutable reuse : bool;
-  (* Whether the thread must be re-added to the pool when the work is
-     done. *)
+type dom = {
+  task_chan : task C.t;
+  mutable domain: unit Domain.t
 }
 
-(* Pool of worker threads: *)
-let workers : thread Queue.t = Queue.create ()
+let workers : dom Queue.t = Queue.create ()
 
-(* Queue of clients waiting for a worker to be available: *)
-let waiters : thread Lwt.u Lwt_sequence.t = Lwt_sequence.create ()
+let waiters : dom Lwt.u Lwt_sequence.t = Lwt_sequence.create ()
 
-(* Code executed by a worker: *)
 let rec worker_loop worker =
-  let id, task = CELL.get worker.task_cell in
-  task ();
-  (* If there is too much threads, exit. This can happen if the user
-     decreased the maximum: *)
-  if !threads_count > !max_threads then worker.reuse <- false;
-  (* Tell the main thread that work is done: *)
-  Lwt_unix.send_notification id;
-  if worker.reuse then worker_loop worker
+  match C.recv worker.task_chan with 
+  | Task (id, task) -> 
+      task ();  
+      Lwt_unix.send_notification id; 
+      worker_loop worker
+  | Quit -> ()
 
-(* create a new worker: *)
 let make_worker () =
-  incr threads_count;
+  incr domains_count;
   let worker = {
-    task_cell = CELL.make ();
-    thread = Thread.self ();
-    reuse = true;
+    task_chan = C.make_bounded 0;
```

Comment:
I'm a little confused since I'm not aware of the internal workings but initializing the channel bound as zero, isn't that equivalent to a synchronous channel? -> https://github.com/ocaml-multicore/domainslib/blob/2ac25c7a207e199281f702560e20f3e7c4069088/lib/chan.ml#L27

Reply:
It is indeed a synchronous channel. Since we are going to send only one task at a time to this channel, shouldn't really matter for performance. If we do see some impact in the performance, we can increase the buffer size.